### PR TITLE
Added no-unnecessary-constructor rule

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### PR checklist

- [x] Addresses an existing issue: #3631
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

Adds a `no-unnecessary-constructor` rule to stop blank constructors.

#### CHANGELOG.md entry:

[new-rule] `no-unnecessary-constructor`

### CODE REVIEW

File: src/rules/noUnnecessaryConstructorRule.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    public static FAILURE_STRING = "There's no need to add an empty constructor to a class.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+const isEmptyConstructor = (node: ts.ConstructorDeclaration): boolean =>
+    node.parameters.length === 0 && node.body !== undefined && node.body.statements.length === 0;
```
Comment (ajafff) [inline src/rules/noUnnecessaryConstructorRule.ts]: IMO only parameter properties should make a constructor necessary. If you need regular parameters for backwards compatibility, you should use a disable comment.

File: test/rules/no-unnecessary-constructor/test.ts.lint [hunk 2]
```diff
@@ -0,0 +1,56 @@
+export class ExportedClass {
```
Comment (ajafff) [inline test/rules/no-unnecessary-constructor/test.ts.lint]: needs tests with overload signatures, and parameter properties
also consider adding a test with an interface containing a construct signature

File: src/rules/noUnnecessaryConstructorRule.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Prevents blank constructors, as they are redundant.",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+            JavaScript implicitly adds a blank constructor when there isn't one.
+            It's not necessary to manually one in.
```
Comment (ajafff) [inline src/rules/noUnnecessaryConstructorRule.ts]: this sentence seems to be missing a verb

File: test/rules/no-unnecessary-constructor/test.ts.lint (hunk 2 above)
Comment (suchanlee) [reply test/rules/no-unnecessary-constructor/test.ts.lint]: +1

File: src/configs/all.ts [hunk 4]
```diff
@@ -125,6 +125,7 @@ export const rules = {
     "no-submodule-imports": true,
     "no-unbound-method": true,
     "no-unnecessary-class": [true, "allow-empty-class"],
+    "no-unnecessary-constructor": [true],
```
Comment (giladgray) [inline src/configs/all.ts]: `true`: no array, like all its siblings

File: src/rules/noUnnecessaryConstructorRule.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        `,
+        ruleName: "no-unnecessary-constructor",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING = "There's no need to add an empty constructor to a class.";
```
Comment (giladgray) [inline src/rules/noUnnecessaryConstructorRule.ts]: this language is too casual.

`Remove unnecessary constructor.`

File: src/rules/noUnnecessaryConstructorRule.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+const isEmptyConstructor = (node: ts.ConstructorDeclaration): boolean =>
+    node.parameters.length === 0 && node.body !== undefined && node.body.statements.length === 0;
+
+const createFix = (node: ts.ConstructorDeclaration): Lint.Replacement =>
+    Lint.Replacement.deleteFromTo(node.pos, node.end);
+
+function walk(context: Lint.WalkContext<void>) {
+    const callback = (node: ts.Node): void => {
+        if (isConstructorDeclaration(node) && isEmptyConstructor(node)) {
+            context.addFailureAtNode(node, Rule.FAILURE_STRING, createFix(node));
```
[the remark this answers is not in the document]
Comment (giladgray) [reply src/rules/noUnnecessaryConstructorRule.ts]: @ajafff how can we resolve your comment? should this rule not have a fixer (sounds potentially unsafe #2625)

**Comment by giladgray** [comment]
@JoshuaKGoldberg please update this branch so we can review it, or close it if no longer relevant. we will close this if we do not hear from you in two weeks.

**Comment by JoshuaKGoldberg** [comment]
Sorry about the delay! I'd missed those comments. They should all be fixed now by adding in extra tests.

**Comment by JoshuaKGoldberg** [comment]
Renamed to `unnecessary-constructor` and verified comments are all addressed, I think.

File: src/rules/noUnnecessaryConstructorRule.ts (hunk 3 above)
Comment (ericanderson) [reply src/rules/noUnnecessaryConstructorRule.ts]: This is still a problem @JoshuaKGoldberg

File: src/rules/unnecessaryConstructorRule.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Prevents blank constructors, as they are redundant.",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+            JavaScript implicitly adds a blank constructor when there isn't one.
+            It's not necessary to manually one in.
```
Comment (ericanderson) [inline src/rules/unnecessaryConstructorRule.ts]: still missing a verb

**Pull request merged**